### PR TITLE
Move tribe to a module

### PR DESCRIPTION
**Pull request by ywelsch**
This PR moves tribe to a module, stripping core from the tribe functionality. If desired, we could even make tribe a plugin instead of a module.

### CODE REVIEW

**Comment by s1monw** [comment]
> If desired, we could even make tribe a plugin instead of a module. 

I think a module is fine here we should have done the move to a plugin earlier we are late in to the game here.

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 1]
```diff
@@ -920,8 +908,23 @@ protected SearchService newSearchService(ClusterService clusterService, IndicesS
         return customNameResolvers;
     }
 
-    /** Constructs an internal node used as a client into a cluster fronted by this tribe node. */
-    protected Node newTribeClientNode(Settings settings, Collection<Class<? extends Plugin>> classpathPlugins, Path configPath) {
+    public static class NodeBuilder {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: do we really need this additional class? it's seem like users can provide all the stuff at once to the newNode method?

File: modules/tribe/src/main/java/org/elasticsearch/tribe/TribeService.java [hunk 2]
```diff
@@ -218,17 +155,31 @@ public static Settings processSettings(Settings settings) {
 
     private final NamedWriteableRegistry namedWriteableRegistry;
 
-    public TribeService(Settings settings, Path configPath, ClusterService clusterService, final String tribeNodeId,
-                        NamedWriteableRegistry namedWriteableRegistry, BiFunction<Settings, Path, Node> clientNodeBuilder) {
+    @Inject
```
Comment (s1monw) [inline modules/tribe/src/main/java/org/elasticsearch/tribe/TribeService.java]: I strongly believe we should not add any new  `@Inject` anywhere in the codebase. we need to find a different way to do this

File: modules/tribe/src/main/java/org/elasticsearch/tribe/TribePlugin.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                                                              NamedWriteableRegistry namedWriteableRegistry,
+                                                              MasterService masterService, ClusterApplier clusterApplier,
+                                                              ClusterSettings clusterSettings, UnicastHostsProvider hostsProvider,
+                                                              AllocationService allocationService) {
+        return Collections.singletonMap("tribe", () -> new TribeDiscovery(settings, transportService, masterService, clusterApplier));
+    }
+
+    @Override
+    public Collection<Class<? extends LifecycleComponent>> getGuiceServiceClasses() {
+        return Collections.singleton(TribeService.class);
```
Comment (s1monw) [inline modules/tribe/src/main/java/org/elasticsearch/tribe/TribePlugin.java]: why do we need to initialize it via guice, is there anything we can do about this? I think this is a blocker if we are adding inject annotations anywhere

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 4]
```diff
@@ -527,6 +517,10 @@ protected Node(final Environment environment, Collection<Class<? extends Plugin>
         }
     }
 
+    public void addOnStartedListener(Runnable runnable) {
+        onStartedListeners.add(runnable);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: our plugin API on the node level is pull based not push. Can we add a method to pull these from `ClusterPlugin.java` that way we are consistent and we fully control who calls this method.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I don't think we need this entire construct we can just subclass Node.java in the plugin / module and create a node via this?

**Review by s1monw — CHANGES_REQUESTED** [review]
I left some suggestions and asks. I love this effort it makes core so much cleaner and settigns handling is so much improved.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/node/Node.java]: To play nicely with the `MockNode` concept used by the tests, we cannot just create the Node instance ourselves. However, by removing Guice injection, I've just replaced this class by a `BiFunction<Settings, Path>`.

File: modules/tribe/src/main/java/org/elasticsearch/tribe/TribeService.java (hunk 2 above)
Comment (ywelsch) [inline modules/tribe/src/main/java/org/elasticsearch/tribe/TribeService.java]: I tried to unguice it in 2025150

File: core/src/main/java/org/elasticsearch/plugins/Plugin.java [hunk 5]
```diff
@@ -107,7 +112,9 @@
      */
     public Collection<Object> createComponents(Client client, ClusterService clusterService, ThreadPool threadPool,
                                                ResourceWatcherService resourceWatcherService, ScriptService scriptService,
-                                               NamedXContentRegistry xContentRegistry) {
+                                               NamedXContentRegistry xContentRegistry, Environment environment,
+                                               NodeEnvironment nodeEnvironment, NamedWriteableRegistry namedWriteableRegistry,
+                                               BiFunction<Settings, Path, Node> clientNodeBuilder) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/plugins/Plugin.java]: I think we should not make this a first class citizen. Instead I suggest we remove the Node#newNode entirely and add a `MockTribePlugin` class that passes such a function that delegates to / creates a `MockNode`. Given that tribe goes away I think that is the right tradeoff. WDYT?

**Review by s1monw — CHANGES_REQUESTED** [review]
Left a comment LGTM otherwise

**Review by s1monw — APPROVED** [review]
LGTM thank you for all the extra iterations. good change!

**Comment by ywelsch** [comment]
Thanks @s1monw

**Pull request merged**